closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java

```diff
@@ -253,116 +234,51 @@ private byte[] implGenerateSeed(int numBytes) {
-                    if (debug != null) {
-                        debug.println("Buffer size changed to " +
-                                new_buffer_size);
-                    }
-                } else {
-                    if (debug != null) {
-                        debug.println("Buffer reached limit: " +
-                                nextBuffer.length);
-                    }
-                }
-                change_buffer = 0;
-            }
-
-            // load fresh random bytes into nextBuffer
-            lastRead = time;
-            readFully(nextIn, nextBuffer);
-            buffered = nextBuffer.length;
+        // always add the seed to our mixing random
+        private synchronized void implSetSeed(byte[] seed) {
+           if (seedOutInitialized == false) {
+               seedOutInitialized = true;
+               seedOut = AccessController.doPrivileged(
+                       new PrivilegedAction<>() {
+                   @Override
+                   public OutputStream run() {
+                       try {
+                           return new FileOutputStream(seedFile, true);
+                       } catch (Exception e) {
+                           return null;
+                       }
+                   }
+               });
+           }
+           if (seedOut != null) {
+               try {
+                   seedOut.write(seed);
+               } catch (IOException e) {
+                   // ignored, on Mac OS X, /dev/urandom can be opened
+                   // for write, but actual write is not permitted
+               }
+           }
+           getMixRandom().engineSetSeed(seed);
         }
 
-        private void implNextBytes(byte[] data) {
+        private synchronized void implNextBytes(byte[] data) {
+            getMixRandom().engineNextBytes(data);
             try {
-                int data_len = data.length;
-                int ofs = 0;
-
-                while (data_len > 0) {
-                    synchronized (LOCK_GET_BYTES) {
-                        ensureBufferValid();
-                        int buf_pos = nextBuffer.length - buffered;
-                        int len;
-
-                        if (data_len > buffered) {
-                            len = buffered;
-                            buffered = 0;
-                        } else {
-                            len = data_len;
-                            buffered -= len;
-                        }
-
-                        System.arraycopy(nextBuffer, buf_pos, data, ofs, len);
-                        ofs += len;
-                        data_len -= len;
-                    }
+                // read random data from non blocking source
+                byte[] rawData = new byte[data.length];
+                readFully(nextIn, rawData);
+                int i = 0;
+                for (byte b : data) {
+                    b ^= rawData[i++];
                 }
```

Comment:
What is the point of this loop?

Reply:
this loop mixes the two random sequences - from /dev/urandom and SHA1PRNG. It ensures that randomness is maintained if /dev/urandom quality decreases due to increased uptake of entropy by the application (as it is not blocking).

Reply:
The loop has no side-effects, there's not even the possibility of an `ArrayIndexOutOfBoundsException`.